modules/caddyhttp/push/handler.go

```diff
@@ -0,0 +1,213 @@
+		err := pusher.Push(repl.ReplaceAll(resource.Target, "."), &http.PushOptions{
+			Method: resource.Method,
+			Header: hdr,
+		})
+		if err != nil {
+			// usually this means either that push is not
+			// supported or concurrent streams are full
+			break
+		}
+	}
+
+	// serve only after pushing!
+	if err := next.ServeHTTP(w, r); err != nil {
+		return err
+	}
+
+	// finally, push any resources described by Link fields that were
+	// written to the response header, only if another push handler
+	// hasn't already done so
+	if links, ok := w.Header()["Link"]; ok {
+		if val := caddyhttp.GetVar(r.Context(), pushedLink); val == nil {
+			h.logger.Debug("pushing Link resources", zap.Strings("linked", links))
+			caddyhttp.SetVar(r.Context(), pushedLink, true)
+			h.servePreloadLinks(pusher, hdr, links)
+		}
+	}
+
+	return nil
+}
+
+func (h Handler) initializePushHeaders(r *http.Request, repl *caddy.Replacer) http.Header {
+	hdr := make(http.Header)
+
+	// prevent recursive pushes
+	hdr.Set(pushHeader, "1")
+
+	// set initial header fields; since exactly how headers should
+	// be implemented for server push is not well-understood, we
+	// are being conservative for now like httpd is:
+	// https://httpd.apache.org/docs/2.4/en/howto/http2.html#push
+	// we only copy some well-known, safe headers that are likely
+	// crucial when requesting certain kinds of content
+	for _, fieldName := range safeHeaders {
+		if vals, ok := r.Header[fieldName]; ok {
+			hdr[fieldName] = vals
+		}
+	}
+
+	// user can customize the push request headers
+	if h.Headers != nil {
+		h.Headers.ApplyTo(hdr, repl)
+	}
+
+	return hdr
+}
+
+// servePreloadLinks parses Link headers from upstream and pushes
+// resources described by them. If a resource has the "nopush"
+// attribute or describes an external entity (meaning, the resource
+// URI includes a scheme), it will not be pushed.
+func (h Handler) servePreloadLinks(pusher http.Pusher, hdr http.Header, resources []string) {
+outer:
+	for _, resource := range resources {
+		for _, resource := range parseLinkHeader(resource) {
+			if _, ok := resource.params["nopush"]; ok {
+				continue
+			}
+			if isRemoteResource(resource.uri) {
+				continue
+			}
+			err := pusher.Push(resource.uri, &http.PushOptions{
+				Header: hdr,
+			})
+			if err != nil {
+				break outer
+			}
+		}
+	}
+}
+
+// Resource represents a request for a resource to push.
+type Resource struct {
+	// Method is the request method, which must be GET or HEAD.
+	// Default is GET.
+	Method string `json:"method,omitempty"`
+
+	// Target is the path to the resource being pushed.
+	Target string `json:"target,omitempty"`
+}
+
+// HeaderConfig configures headers for synthetic push requests.
+type HeaderConfig struct {
+	headers.HeaderOps
+}
+
+// isRemoteResource returns true if resource starts with
+// a scheme or is a protocol-relative URI.
+func isRemoteResource(resource string) bool {
+	return strings.HasPrefix(resource, "//") ||
```

Comment:
While it’s probably good enough for now, the spec allows a server to push any resource it has authority (as defined in the HTTPS spec) on. Basically, a server can push resources for any domain present in its TLS certificate. This probably require a new config option to be handled properly.

Reply:
Yeah, this is complicated to get 100% correct. But I don't even know if the Go standard lib works well in that case, since the push requests are all synthetic and within the same Server instance; i.e. we can't push resources handled by external servers (without some sort of proxy configured separately within the current server).

I'm hoping/expecting that it's a very rare edge case anyway.